### PR TITLE
fix url encoding

### LINKED ISSUE #799: list in query params encoding url bug
### Issue Info

| Info                           | Value                                                 |      
| ------------------------------ | ----------------------------------------------------- | 
| Platform Name                  | flutter                      |     
| Platform Version               | 1.17.1                            |     
| Dio Version                    |3.0.9                                |     
| Repro rate                     |  all the time (100%)  |    

### Issue Description and Steps

When I send request with list of enums in queryParams 

(for example {'a' :1,'b': 2, color:[Colors.red, Colors.white]})

in URI query params i get "a=1&b=2&**color[]=RED&color[]=WHITE**"
the square brackets in key name isn't good
I think that urlEncode in utils.dart doesn't work properly

### PR DESCRIPTION
**Pull request by julek-kal**
### New Pull Request Checklist

- [x] I have read the [Documentation](https://pub.dartlang.org/packages/dio)
- [x] I have searched for a similar pull request in the [project](https://github.com/flutterchina/dio/pulls) and found none
- [x] I have updated this branch with the latest master to avoid conflicts (via merge from master or rebase)
- [x] I have added the required tests to prove the fix/feature I am adding
- [x] I have updated the documentation (if necessary)
- [x] I have run the tests and they pass

This merge request fixes / refers to the following issues: 
Fixes #799 

### Pull Request Description
Added collection format like in [swagger docs](https://swagger.io/docs/specification/2-0/describing-parameters/#array-and-multi-value-parameters)

default CollectionFormat is csv
...

### CODE REVIEW

**Comment by frameset** [comment]
Ignoring it doesn't make it stale Mr Bot.

**Comment by josh-burton** [comment]
Have hit this bug as well. Would love to see this merged.

**Comment by josh-burton** [comment]
Any update? LGTM!

File: dio/lib/src/options.dart [hunk 1]
```diff
@@ -40,6 +42,27 @@ enum ResponseType {
...
+enum CollectionFormat {
+  // Comma-separated values
+  // e.g. (foo,bar,baz)
+  csv,
+  // Space-separated values
+  // e.g. (foo bar baz)
+  ssv,
+  // Space-separated values
+  // e.g. (foo\tbar\tbaz)
+  tsv,
```
Comment (lyio) [inline dio/lib/src/options.dart:57]: are these comments correct?

Comment (julek-kal) [reply dio/lib/src/options.dart:57]: No, it's my mistake. It's should be "Tab-separated values"

File: dio/lib/src/options.dart [hunk 2]
```diff
@@ -40,6 +42,27 @@ enum ResponseType {
...
+// (a single parameter with multiple parameter or multiple parameters with the same name)
+// and the separator for array items.
+enum CollectionFormat {
+  // Comma-separated values
+  // e.g. (foo,bar,baz)
+  csv,
+  // Space-separated values
+  // e.g. (foo bar baz)
+  ssv,
+  // Space-separated values
```
Comment (julek-kal) [inline dio/lib/src/options.dart:55]:
```suggestion
  // Tab-separated values
```

File: dio/lib/src/options.dart [hunk 3]
```diff
@@ -40,6 +42,27 @@ enum ResponseType {
...
+  // Comma-separated values
+  // e.g. (foo,bar,baz)
+  csv,
+  // Space-separated values
+  // e.g. (foo bar baz)
+  ssv,
+  // Tab-separated values
+  // e.g. (foo\tbar\tbaz)
+  tsv,
+  // Space-separated values
```
Comment (julek-kal) [inline dio/lib/src/options.dart:58]:
```suggestion
  // Pipe-separated values
```

File: dio/lib/src/options.dart [hunk 4]
```diff
@@ -40,6 +42,27 @@ enum ResponseType {
...
+  // e.g. (foo\tbar\tbaz)
```
Comment (julek-kal) [inline dio/lib/src/options.dart:59]:
```suggestion
  // e.g. (foo|bar|baz)
```

File: dio/lib/src/options.dart (hunk 1 above)
Comment (julek-kal) [reply dio/lib/src/options.dart:57]: It's fixed

**Comment by kuhnroyal** [comment]
@wendux When can we expect a new dio release with this fix/feature? I would like to implement this in the openapi-generator.

**Comment by praveenkutta08** [comment]
Has this issue been resolved?

**Comment by frameset** [comment]
Please consider merging this PR, it solves a critical issue for several of us.

**Comment by kuhnroyal** [comment]
@julek-kal Can you rebase this please

**Comment by julek-kal** [comment]
@kuhnroyal of course, but now I can't. I'll do it tomorrow

**Comment by julek-kal** [comment]
@kuhnroyal Theoretically i rebased this, but i can't run tests because of #914 so shouldn't push it 😞

**Comment by kuhnroyal** [comment]
You probably need to change the pubspecs all to the path versions via `dependency_overrides`. Not sure.

**Comment by julek-kal** [comment]
Ok, I cloned original dio now for certainty that errors doesn't follow from mistakes during rebase. I changed pubspecs(pub get start working), even after that i'm getting 126 errors, related to non nullability especially(tests still doesn't execute)

**Comment by julek-kal** [comment]
@kuhnroyal if you want I can push rebased version to another branch and merge it when I will run tests successfully

**Comment by kuhnroyal** [comment]
Yea, I think master is broken. Merged old PRs after the null-safety PR.

**Comment by kuhnroyal** [comment]
Looks like @wendux merged this or added some other collection format handling.

**Comment by julek-kal** [comment]
Right, so I guess it's can be closed, but I will leave it to @wendux decision

**Comment by wendux** [comment]
I have merge this feature in 4.0 branch ( beta )

**Comment by kuhnroyal** [comment]
Thank you @wendux!

I tried to start implementing this in the OpenAPI generator and noticed that the format may be different per parameter (form and/or query parameter).

So I am think if we can allow `queryParameters` and `x-www-url-encoded` body to alternatively accept some `Param` object.
Basically something like this:

```dart
final response = await dio.request<Object>(
      '/foo',
      data: <String, dynamic>{
         'value': 7,
         'value2': Param<int>(3),
         'tags': MultiParam<String>(['foo', 'bar'], format: ListFormat.csv),
         'list_with_default_format': [1, 2, 3],
      },
      queryParameters: <String, dynamic>{
         'value': 7,
         'value2': Param<int>(3),
         'tags': MultiParam<String>(['foo', 'bar'], format: ListFormat.pipes),
      },
    );
```

What do you guys think? Is that possible?

**Comment by wendux** [comment]
> Thank you @wendux!
> 
> I tried to start implementing this in the OpenAPI generator and noticed that the format may be different per parameter (form and/or query parameter).
> 
> So I am think if we can allow `queryParameters` and `x-www-url-encoded` body to alternatively accept some `Param` object.
> Basically something like this:
> 
> ```dart
> final response = await dio.request<Object>(
>       '/foo',
>       data: <String, dynamic>{
>          'value': 7,
>          'value2': Param<int>(3),
>          'tags': MultiParam<String>(['foo', 'bar'], format: ListFormat.csv),
>          'list_with_default_format': [1, 2, 3],
>       },
>       queryParameters: <String, dynamic>{
>          'value': 7,
>          'value2': Param<int>(3),
>          'tags': MultiParam<String>(['foo', 'bar'], format: ListFormat.pipes),
>       },
>     );
> ```
> 
> What do you guys think? Is that possible?

This may be complicated and not very general...

**Comment by kuhnroyal** [comment]
I agree that it is unlikely but I am trying to generate code that is compliant with https://swagger.io/specification/#parameter-style which allows this to be different for each query parameter.

I made a simple draft for discussion here: https://github.com/flutterchina/dio/pull/1063

**Comment by kuhnroyal** [comment]
I think this PR can be closed, further discussion should move back to the issue. Was my bad posting my other comments here.

**Pull request closed without merging**